rust/lance-encoding/src/encodings/logical/primitive.rs

```diff
@@ -4117,31 +4121,150 @@ impl PrimitiveStructuralEncoder {
+            .get_stat(Stat::Cardinality)?
+            .as_primitive::<UInt64Type>()
+            .value(0);
+
+        let num_values = data_block.num_values();
+
+        match data_block {
+            DataBlock::FixedWidth(_) => {
+                // Dictionary: cardinality unique values at 128 bits each
+                let dict_size = cardinality * (DICT_FIXED_WIDTH_BITS_PER_VALUE / 8);
+                // Indices: num_values indices at 32 bits each
+                let indices_size = num_values * (DICT_INDICES_BITS_PER_VALUE / 8);
+                Some(dict_size + indices_size)
+            }
+            DataBlock::VariableWidth(var) => {
+                // Only 32-bit and 64-bit offsets are supported
+                if var.bits_per_offset != 32 && var.bits_per_offset != 64 {
+                    return None;
+                }
+                let bits_per_offset = var.bits_per_offset as u64;
+
+                let data_size = data_block
+                    .get_stat(Stat::DataSize)?
+                    .as_primitive::<UInt64Type>()
+                    .value(0);
+                let avg_value_size = data_size / num_values;
+
+                // Dictionary values: actual bytes of unique strings/binary
+                let dict_values_size = cardinality * avg_value_size;
+                // Dictionary offsets: pointers into dictionary values
+                let dict_offsets_size = cardinality * (bits_per_offset / 8);
+                // Indices: map each row to dictionary entry
+                let indices_size = num_values * (bits_per_offset / 8);
+
+                Some(dict_values_size + dict_offsets_size + indices_size)
+            }
+            _ => None,
+        }
+    }
+
+    fn should_dictionary_encode_based_on_size(data_block: &DataBlock, field: &Field) -> bool {
+        // Get size ratio from metadata or env var, default to 0.8
+        let threshold_ratio = field
+            .metadata
+            .get(DICT_SIZE_RATIO_META_KEY)
+            .and_then(|val| val.parse::<f64>().ok())
+            .or_else(|| {
+                env::var("LANCE_ENCODING_DICT_SIZE_RATIO")
+                    .ok()
+                    .and_then(|val| val.parse().ok())
+            })
+            .unwrap_or(0.8);
+
+        // Validate size ratio is in valid range
+        if threshold_ratio <= 0.0 || threshold_ratio > 1.0 {
+            panic!(
+                "Invalid parameter: dict-size-ratio is {} which is not in the range (0, 1].",
+                threshold_ratio
+            );
         }
 
-        // Somewhat arbitrary threshold rule.  Apply dictionary encoding if the number of unique
-        // values is less than 1/2 the total number of values.
-        let divisor: u64 = field
+        // Get raw data size
+        let data_size = if let Some(data_size_array) = data_block.get_stat(Stat::DataSize) {
+            data_size_array.as_primitive::<UInt64Type>().value(0)
+        } else {
+            return false;
+        };
```

Comment:
I didn't actually remember that we have a stat named `Stat::DataSize`.  Can we just use `data_block.data_size()`?  That is simpler, though maybe it's not exactly the same thing?